frame/staking/src/lib.rs

```diff
@@ -1431,17 +1481,21 @@ decl_module! {
 				ledger = ledger.consolidate_unlocked(current_era)
 			}
 
-			if ledger.unlocking.is_empty() && ledger.active.is_zero() {
+			let post_info_weight = if ledger.unlocking.is_empty() && ledger.active.is_zero() {
 				// This account must have called `unbond()` with some value that caused the active
 				// portion to fall below existential deposit + will have no more unlocking chunks
 				// left. We can now safely remove all staking-related information.
-				Self::kill_stash(&stash)?;
+				Self::kill_stash(&stash, num_slashing_spans)?;
 				// remove the lock.
 				T::Currency::remove_lock(STAKING_ID, &stash);
+				// This is worst case scenario, so we use the full weight and return None
+				None
 			} else {
 				// This was the consequence of a partial unbond. just update the ledger and move on.
 				Self::update_ledger(&controller, &ledger);
-			}
+				// This is only an update, so we use less overall weight
+				Some(50 * WEIGHT_PER_MICROS + T::DbWeight::get().reads_writes(4, 2))
+			};
```

Comment:
This is very neat, but what is the rational between choosing this approach and just asking the user to submit en extra bit of bool that would indicate if this would cause an update or kill. I think for the chain it is slightly cheaper to do this. In some cases in elections-phragmen, I chose this approach. I hope both are fine.

Reply:
for now I always choose something which is not a breaking change over something that is a breaking change.

Reply:
I think it depends on how much of a change it would make. I don't know if benchmarks figured it out but at least in this case I can intuitively say it is not significant. 

Otherwise, if the change is significant, we must do a breaking change, as we already have in numerous extrinsics.

Reply:
It would be better imo to split this into two extrinsics rather than add a bool if we were going down the path of introducing breaking changes

Reply:
for example, the whole slashing span this is not even used in this arm of the logic